8269907 memory leak - Dirty Nodes / Parent removed

FlorianKirmaier opened the pull request:
After thinking about this issue for some time, I've now got a solution.
I know put the scene in the state it is, before is was shown, when the dirtyNodes are unset, the whole scene is basically considered dirty. 
This has the drawback of rerendering, whenever a window is "reshown", but it restores sanity about memory behaviour, which should be considered more important.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8269907](https://bugs.openjdk.org/browse/JDK-8269907)

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8269907: memory leak - Dirty Nodes / Parent removed`

### Issue
 * [JDK-8269907](https://bugs.openjdk.org/browse/JDK-8269907): MemoryLeak: Scene.dirtyNodes and Parent.removed referencing removed Nodes when window is not showing (**Bug** - P4) ⚠️ Title mismatch between PR and JBS.


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/584/head:pull/584` \
`$ git checkout pull/584`

Update a local copy of the PR: \
`$ git checkout pull/584` \
`$ git pull https://git.openjdk.org/jfx.git pull/584/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 584`

View PR using the GUI difftool: \
`$ git pr show -t 584`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/584.diff">https://git.openjdk.org/jfx/pull/584.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/584#issuecomment-884120569)

kevinrushforth commented:
/reviewers 2

kevinrushforth reviewed (changes requested):
This proposed fix will almost certainly lead to threading problems. You can't call `synchronizeSceneNodes` or `syncAll` at an arbitrary time, but only during the pulse callback, which is called with the `renderLock` held.

I also see unit test failures.

FlorianKirmaier commented:
Ok, thank you for the feedback.

Is there a way, to do something with the renderlock on, with an similar API as Platform.runLater? 

Then it would also be possible, to clear the dirtyNodes in the supposed way, without causing the minor performance issue of rerendering the whole Scene again. This would be a much cleaner approach.

FlorianKirmaier commented:
During the past PRs the unit tests were quite unstable, but they seem to work.

I've reworked my fix. It now registers a pulseListener for the cleanup. The is also how i initially thought, the fix should look like. To do so, I've added the option to the Toolkit to register such listeners. 

The fix no longer has the drawback, of rerendering the whole window, when it's showing property is changed.

kevinrushforth commented:
The revised approach doesn't appear to have the same threading issues that the first attempt did (good). I would have thought you could add some logic to the existing scene pulse listener instead of creating a new one. I'll take a closer look when I review this.

FlorianKirmaier commented:
It would be possible to add it to the existing PulseListener, but i think it would be more complicated. It might require some code to decide whether it should only clean the dirty children, or do the full pulse. The logic to add/remove the listener also would have to be more complicated.

Actually it was my first approach, but i didn't find a good place "to get started".

kevinrushforth commented:
I took a closer look, and the current PR has at least one problem. The call to `synchronizeSceneNode` must only be done with the `renderLock` held. If you are using this in production, then it likely means you aren't hitting the cases where this would break. I recommend either my earlier suggestion of hooking into the existing listener, or else you will need to acquire the `renderLock`. If the latter, then one thing you will want to avoid is running both a regular sync pass and a cleanup pass in the same pulse. I see that you try to ensure that by only adding the listener if the scene is not showing, but since you might have to also check in the listener itself (not sure).

FlorianKirmaier commented:
Thank you for your feedback!
To solidify my understanding - the "RenderScenegraph" is only allowed to be changed when the render lock is held. This scene graph is represented by all these NG<Something> classes, correct?

I've now updated my PR. I found some code for snapshot, which solves the same problem about aquiring the render lock. I've moved it into an own method and used it for my fix.

The regular sync, and the new sync shouldn't happen in the same pulse - because the regular sync checks whether the Window is closed by `if (peer != null)`. At least it shouldn't require the lock. If it would for some reason aquire the lock, then it should basically be a noop - so i don't really see why this would be an issue. Performance wise aquiring an additional lock can cause a very minor slowdown, but because it only happens once per closed window, it is realy minor.

kevinrushforth commented:
> To solidify my understanding - the "RenderScenegraph" is only allowed to be changed when the render lock is held. This scene graph is represented by all these NG classes, correct?

Correct. When the renderLock is not held, the renderer thread is free to render it and can be sure it  won't change.

I'll take a look at the updated fix.

FlorianKirmaier commented:
If found a bug in the current implementation and fixed it - which took quite long.
It seems like, calling synchronizeSceneNodes without calling updateBounds() can cause that the NGNode can get out of sync. Which then caused rendering bugs with disappearing objects.

This is now fixed. 

I've also added a call to the `syncLights` and `synchronizeSceneProperties` methods.
I think it's less complex to just sync everything, instead of just syncing a subset - also when making a snapshot.

Maybe it would make sense to add a check like the following, to the doUpdatePeer() method?
```
            if(txBoundsInvalid) {
                    PlatformLogger logger = Logging.getInputLogger();
                    if (logger.isLoggable(Level.WARN)) {
                        logger.finer("txBoundsInvalid ot of sync on doUpdatePeer");
                    }
            }
```

christianheilmann commented:
How to proceed to get this PR reviewed?

FlorianKirmaier commented:
I've just merged it with master, to make it easier to review, if someone finds the time.

FlorianKirmaier commented:
I've just merged with master again.
It would be great if this could be reviewed at some time.

mstr2 commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java:
```diff
@@ -414,11 +416,15 @@ public void firePulse() {
         final Map<TKPulseListener,AccessControlContext> postScenePulseList =
                 new WeakHashMap<>();
+        final Map<TKPulseListener,AccessControlContext> cleanupList =
+                new WeakHashMap<>();
```
I wonder why we need `WeakHashMap` here. These maps are short-lived anyway, since they're local variables.

FlorianKirmaier replied:
I've changed it now to a normal HashMap.

mstr2 commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java:
```diff
@@ -480,6 +489,11 @@ public void removePostSceneTkPulseListener(TKPulseListener listener) {
 
+    public void addCleanupListener(TKPulseListener listener) {
+        AccessControlContext acc = AccessController.getContext();
+        cleanupListeners.put(listener,acc);
```
Access to `cleanupListeners` is not synchronized on `this` here, but it is synchronized in L426.
You should either synchronize each and every access, or none of it.

FlorianKirmaier replied:
I've added a synchronized now.

mstr2 commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -732,6 +734,21 @@ public final void removePostLayoutPulseListener(Runnable r) {
+    };
+    private void checkCleanDirtyNodes() {
+        if(!cleanupAdded) {
+            if((window.get() == null || !window.get().isShowing()) && dirtyNodesSize > 0) {
```
Minor: space after `if`.
You could also reduce nesting by consolidating the two separate conditions.

mstr2 commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -2090,19 +2109,24 @@ null, new Point3D(
      **************************************************************************/
 
-    private void windowForSceneChanged(Window oldWindow, Window newWindow) {
+    private void windowForSceneChanged(Window oldWindow, Window window) {
```
This change doesn't seem necessary.

FlorianKirmaier replied:
I've reverted the variable name.

mstr2 commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -2090,19 +2109,24 @@ null, new Point3D(
+        checkCleanDirtyNodes();
     }
 
+    private final ChangeListener<Boolean> sceneWindowShowingListener = (p, o, n) -> {checkCleanDirtyNodes(); } ;
```
Minor: should be no space before `;`.
You could also remove the curly braces.

FlorianKirmaier replied:
I've removed now the curly braces.

mstr2 commented on tests/system/src/test/java/test/javafx/scene/DirtyNodesLeakTest.java:
```diff
@@ -0,0 +1,84 @@
+import javafx.stage.Stage;
+import junit.framework.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
```
Since this is a new class, you should use the JUnit5 API.

FlorianKirmaier replied:
I'm now using the junit5 api.

mstr2 commented on tests/system/src/test/java/test/javafx/scene/DirtyNodesLeakTest.java:
```diff
@@ -0,0 +1,84 @@
+
+import static test.util.Util.TIMEOUT;
+
+public class DirtyNodesLeakTest {
```
Since this tests dirty nodes of a `Scene`, maybe you could use a name like `Scene_dirtyNodesLeakTest`?

FlorianKirmaier replied:
I like the name you've suggested. I've changed it now to your suggestion.

mstr2 reviewed (commented):
I've left come comments.

hjohn commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -732,6 +734,21 @@ public final void removePostLayoutPulseListener(Runnable r) {
+        // JDK-8269907 - This is important, to avoid memoryleaks in dirtyNodes and Parent.removed
+        synchronizeSceneNodesWithLock();
+    };
+    private void checkCleanDirtyNodes() {
```
minor: empty line to separate methods and field declarations

hjohn commented on tests/system/src/test/java/test/javafx/scene/DirtyNodesLeakTest.java:
```diff
@@ -0,0 +1,84 @@
+
+                checker.setAsReferenced(scene);
+                checker.assertCollectable(label);
+                stage.setOnShown(l -> {
+                    Platform.runLater(() -> {
+                        root.getChildren().clear();
+                        stage.close();
+                        showingLatch.countDown();
+                    });
+                });
```
I've been looking at this solution, and from what I see, we're doing our best to add a clean up listener which (as far as I can tell without any comments being added) is called when the stage is no longer shown.

The test above creates a stage, adds a label to it in a stack pane, then clears the children.  I think that should be enough to make the label collectable.

I get the impression that you also have to close the stage to make it collectable, which should not be needed at all.

If I'm correct, then this solution introduces quite some new code but doesn't properly address the issue. The dirty nodes list is clearly holding on to nodes that may not be part of the scene any more. 

Looking at the code, I see that `dirtyNodes` is a poor copy of what `ArrayList` does.  It doesn't clear the array with `null`s when it is resizing the array or when it removes elements from the array (or clears it).

The code that I think is the culprit is this part:

                // This is not the first time this scene has been synchronized,
                // so we will only synchronize those nodes that need it
                for (int i = 0 ; i < dirtyNodesSize; ++i) {
                    Node node = dirtyNodes[i];
                    dirtyNodes[i] = null;
                    if (node.getScene() == Scene.this) {
                            node.syncPeer();
                    }
                }
                dirtyNodesSize = 0;

The problem here is that it sets `dirtyNodesSize` to 0, but doesn't clear the array, leaving references to all kinds of nodes in there.

My recommendation to solve this would be to change this code to use `ArrayList` instead of a hand rolled solution. `ArrayList` will properly `null` out slots in its array when they become unused.

FlorianKirmaier replied:
> The test above creates a stage, adds a label to it in a stack pane, then clears the children. I think that should be enough to make the label collectable.
Thats correct. In this case it is collectable.

But the point of the test is, to make sure it's also collectable when the stage no longer renders.

I've just looked at it - and to my impression, everything works as expected. 
It's collectable when the stage is showing.
But not collectable when the stage gets immediately closed (which is the bug of this PR)

hjohn reviewed (changes requested):
I think this can be fixed differently and with far less changes required.

See the comment I added on the test case.

hjohn commented:
I took another good look at this solution, and I would like to offer an alternative as I think this solution is more dealing with symptoms of the underlying problem.

I think the underlying problem is:

1) `Parent` is tracking a `removed` list, which is a list of `Node`. However, it only requires the peer for the dirty area calculation. This is a classic case of not being specific enough with the information you need. If `Parent` is modified to store `NGNode` in its `removed` list, then the problem of `removed` keeping references to old nodes disappears as `NGNode` does not refer back to `Node`.  Note: there is a test method currently in `Parent` that returns the list of removed nodes -- these tests would need to be modified to work with a list of `NGNode` or some other way should be found to check these cases.

2) `Scene` keeps tracking dirty nodes even when it is not visible. The list of dirty nodes could (before this fix) become as big as you want as it was never cleared as the pulse listener that synchronizes the nodes never runs when `peer == null` -- keep adding and removing new nodes while the scene is not shown, and the list of dirty nodes grows indefinitely. This only happens if the scene has been shown at least once before, as before that time special code kicks in that tries to avoid keeping track of all scene nodes in the dirty list.

I think the better solution here would be to reset the scene to go back to its initial state (before it was shown) and sync all nodes when it becomes visible again. This can be achieved by setting the dirty node list to `null` to trigger the logic that normally only triggers on the first show.  `addToDirtyList` should simply do nothing when `peer == null` -- with the current solution, peers are still being updated (in other words, hiding the window and then changing controls still updates peers).

I believe the `syncPeer` call is smart enough not to update the peer in the situation where a scene is hidden and then reshown, even if `Scene` calls it again on all its nodes (`syncPeer` in `Node` will check `dirtyBits.isEmpty()`).

FlorianKirmaier commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -732,6 +734,21 @@ public final void removePostLayoutPulseListener(Runnable r) {
+    private void checkCleanDirtyNodes() {
+        if(!cleanupAdded) {
+            if((window.get() == null || !window.get().isShowing()) && dirtyNodesSize > 0) {
+                Toolkit.getToolkit().addCleanupListener(cleanupListener);
```
I assume that these methods are just called all the time - independent of whether there is an open window.
I would like to leave it like this - unless there is an issue and an idea on how to do it differently.

FlorianKirmaier commented:
@hjohn 
```
The Parent is tracking a removed list, which is a list of Node. However, it only requires the peer for the dirty area calculation. This is a classic case of not being specific enough with the information you need. If Parent is modified to store NGNode in its removed list, then the problem of removed keeping references to old nodes disappears as NGNode does not refer back to Node. Note: there is a test method currently in Parent that returns the list of removed nodes -- these tests would need to be modified to work with a list of NGNode or some other way should be found to check these cases.
```
This would just end in an leak of the NGNode, instead of the Node?
There are also some leaks related to NGNode, bug guess I'll have to fix the more obvious bugs first.
Also, NGNode sometimes refers to the Node. There are a lot of issues in this area.


## Resetting the Scene
This approach has one problem - it would change the complexity of showing/hiding a window. This issue typically happened with Popups, which are regularly shown/hidden. I guess changing the complexity to O(<number-of-node>) might be ok - but I'm not sure about it.

I don't want to change the dirtyNodes to ArrayList in this PR, i don't want to mix this kind of refactoring into the Bugfix-PR.

FlorianKirmaier commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java:
```diff
@@ -480,6 +489,11 @@ public void removePostSceneTkPulseListener(TKPulseListener listener) {
         }
     }
 
+    public void addCleanupListener(TKPulseListener listener) {
```
I've changed it to addCleanupTask

hjohn commented:
> @hjohn
> 
> ```
> The Parent is tracking a removed list, which is a list of Node. However, it only requires the peer for the dirty area calculation. This is a classic case of not being specific enough with the information you need. If Parent is modified to store NGNode in its removed list, then the problem of removed keeping references to old nodes disappears as NGNode does not refer back to Node. Note: there is a test method currently in Parent that returns the list of removed nodes -- these tests would need to be modified to work with a list of NGNode or some other way should be found to check these cases.
> ```
> 
> This would just end in an leak of the NGNode, instead of the Node? There are also some leaks related to NGNode, bug guess I'll have to fix the more obvious bugs first. Also, NGNode sometimes refers to the Node. There are a lot of issues in this area.

Why not fix the root causes? This PR introduces several new things to fix a bug in a round about way that would not be needed at all if the root causes are fixed. I'm happy to help out here, as I prefer fixing the underlying problems (which probably solves multiple problems at once) rather than having to deal with each symptom and making things even more inscrutable in places where these symptoms appear.

If you know the root causes, then please explain them so they can be dealt with. I've only taken a short look, and I agree that `NGNode` probably would be leaked instead. However, the end goal is to track how big the dirty area is. If keeping a list of nodes/ngnodes is bad, then how about keeping a rectangle instead and tracking the dirty area immediately?  That would also solve problems related to NGNode pointing back to Node.

This would likely solve multiple problems, not just related to stages -- I suppose nodes can be detached and this removed list could cause problems then as well.
 
> ## Resetting the Scene
> This approach has one problem - it would change the complexity of showing/hiding a window. This issue typically happened with Popups, which are regularly shown/hidden. I guess changing the complexity to O() might be ok - but I'm not sure about it.

I think that the resulting code would be simpler, with less special cases (why should a stage that was shown before act differently from a fresh one). The O(n) loop to sync all peers should not be an issue when doing a relatively heavy action like showing a window as `syncPeer` will do nothing for nodes that haven't become dirty.

Also, it has benefits as well. Changes to nodes on a shown then hidden stage would now be more performant, and would eliminate any performance differences between a fresh stage and one that was shown before.

hjohn commented:
Adding to the above, I couldn't find in what cases `Node` might be referred from `NGNode`. Where is this happening?

Further, `NGNode` only has a parent pointer, not a full parent/child hierarchy like `Parent` has.  This would mean that if you track `NGNode` in `Parent#removed` a big part of the problem is alleviated as only a few parent `NGNode` will be unable to be GC'd, versus entire UI trees when you refer to `Node`.

I also did some more digging, and the only thing that the dirty region computation for removed nodes really requires is a `RectBounds`.  To do this it accesses only a couple of fields of `NGNode`, no heavy calculations are involved in this part and so this could be done right away -- it's also possible that `Node` may already have this information in some form (basically we need to know the last bounds of the node when it was rendered on screen -- saving a copy in the `doUpdatePeer` call may be sufficient).  When the dirty area computation for the parent is performed, you'd only need to apply the transforms and clipping logic that the parent provides on this `RectBounds` -- no further fields of the removed `NGNode` would need to be accessed.

FlorianKirmaier commented on modules/javafx.graphics/src/main/java/javafx/scene/Scene.java:
```diff
@@ -1281,20 +1298,22 @@ void doCSSLayoutSyncForSnapshot(Node node) {
+        synchronizeSceneNodesWithLock();
+    }
+
+    private void synchronizeSceneNodesWithLock() {
```
I've changed it to synchronizeSceneNodes quite some time ago.

FlorianKirmaier commented:
@hjohn 
Sorry for the late reply.

> Further, NGNode only has a parent pointer, not a full parent/child hierarchy like Parent has. 
An NGNode is usually an NGGroup, which has its full tree referenced.

I guess it would be possible, to do the following changes:
 - Always rerender a stage, if it's shown again - instead of only rendering the dirty area (and remove dirtyNodes when the stage is no longer shown)
 - Remove the "removed" nodes from NGGroup. (either by refactoring and precalculating the dirty area, or just clearing it when a window is no longer shown)
This would change the O(n) complexity of showing/hiding a stage, but the rendering would be less complex.
Imo, this would require more intrusive changes than this PR.
I guess someone has to decide on it.

What I don't understand is, why this PR is considered to be complex.
It just adds some kind of runLater to the Toolkit (I'm surprised, it didn't already exist) and calls it to clean up the dirtyNodes, when a window is no longer shown. I think that's quite straightforward.

FlorianKirmaier commented:
/open

I merged with master.

FlorianKirmaier commented:
Worth mentioning, that I've been using this version of the fix quite some long, and it works well with any issues.
Of course another solution might be good too.

FlorianKirmaier commented:
When we recompute the dirty Region, directly when a node is removed,
then we also change a bit the performance behaviour.
Because it's often the case, that only the structure, of the scene-graph is rebuilt.
So making the suggested changes also might change the performance behaviour.

The version of this PR is relative minimal and not very invasive to the current logic.
